ktor-server-rate-limiting/src/commonMain/kotlin/io/github/flaxoos/ktor/server/plugins/ratelimiter/implementations/Bucket.kt

```diff
@@ -0,0 +1,143 @@
+        by: Double,
+        shouldUpdateTime: Boolean = false,
+        consideringLastUpdateTime: suspend Double.(Long) -> Double = { this }
+    ): Double? {
+        by.assertNotNegative()
+        return increaseVolume(call, by, shouldUpdateTime, consideringLastUpdateTime)
+    }
+
+    /**
+     * Increases the volume by the given amount, coercing to at most the capacity
+     *
+     * @param shouldUpdateTime whether to update the last update time
+     * @param by the amount to increase the volume by, must be positive
+     */
+    internal suspend fun increaseVolume(
+        call: ApplicationCall,
+        by: Double,
+        shouldUpdateTime: Boolean = false,
+        consideringLastUpdateTime: suspend Double.(Long) -> Double = { this }
+    ): Double? {
+        by.assertNotNegative()
+        return tryUpdateVolume(call, shouldUpdateTime) { volume, timeSinceLastUpdate ->
+            volume + (consideringLastUpdateTime(by, timeSinceLastUpdate)).coerceAtMost(capacity.toDouble())
+        }
+    }
+
+    /**
+     * Tries to update the volume if the updated volume is within the capacity of the bucket
+     *
+     * @param update the function to update the volume, with the current volume and the time since the last update as
+     * the arguments
+     * @return the updated volume if it is within the capacity of the bucket,
+     * or null if the updated volume is not within the capacity
+     */
+    private suspend fun tryUpdateVolume(
+        call: ApplicationCall,
+        shouldUpdateTime: Boolean = false,
+        update: suspend (Double, Long) -> Double
+    ): Double? {
+        return currentVolumeMutex.withLock {
+            val now = clock()
+            val lastUpdateTime = _lastUpdateTime.value
+            val timeSinceLastUpdate = now - lastUpdateTime
+            log.trace {
+                "${call.id()}: Maybe updating volume at now: ${fromEpochMilliseconds(now)}, " +
+                    "lastUpdateTime: ${fromEpochMilliseconds(lastUpdateTime)}, " +
+                    "diff: $timeSinceLastUpdate"
+            }
+
+            val updatedVolume = update(currentVolume, timeSinceLastUpdate)
+            val diff = updatedVolume - currentVolume
+
+            if (
+                0 <= updatedVolume &&
+                updatedVolume <= capacity * callVolumeUnit.size
+            ) {
+                log.trace { "${call.id()}: Updating volume from $currentVolume to $updatedVolume" }
+                currentVolume = updatedVolume
+                if (shouldUpdateTime) {
+                    _lastUpdateTime.update { now }
+                }
+                currentVolume
+            } else {
+                log.trace { "${call.id()}: Volume not updated due to exceeding volume: currentVolume: $currentVolume, updatedVolume: $updatedVolume, diff: $diff, max: ${capacity * callVolumeUnit.size}" }
+                null
+            }
+        }
+    }
+
+    private fun Double.assertNotNegative(): Double {
```

Comment:
I think this is a bit over the top. It's just misdirection. Also assert has test connotations.

Reply:
What do you mean by misdirection? the reason there's a function here is because that condition must be met in different places, so avoiding duplication.
Ok about the "assert" changing to "check"

Reply:
A stack trace will lead to this method